src/cargo/core/compiler/output_sbom.rs

```diff
@@ -0,0 +1,292 @@
+use cargo_util_schemas::core::PackageIdSpec;
+use itertools::Itertools;
+use semver::Version;
+use serde::Serialize;
+
+use crate::core::profiles::{DebugInfo, Lto, Profile};
+use crate::core::{Target, TargetKind};
+use crate::util::Rustc;
+use crate::CargoResult;
+
+use super::{unit_graph::UnitDep, BuildRunner, CrateType, Unit};
+
+#[derive(Serialize, Clone, Debug, Copy)]
+#[serde(rename_all = "kebab-case")]
+enum SbomBuildType {
+    /// A package dependency
+    Normal,
+    /// A build script dependency
+    Build,
+}
+
+/// Typed version of a SBOM format version number.
+pub struct SbomFormatVersion<const V: u32>;
+
+impl<const V: u32> Serialize for SbomFormatVersion<V> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_u32(V)
+    }
+}
+
+/// A profile can be overriden for individual packages.
+///
+/// This wraps a [`Profile`] object.
+/// See <https://doc.rust-lang.org/nightly/cargo/reference/profiles.html#overrides>
+#[derive(Serialize, Clone, Debug)]
+struct SbomProfile {
+    name: String,
+    opt_level: String,
+    lto: Lto,
+    codegen_backend: Option<String>,
+    codegen_units: Option<u32>,
+    debuginfo: DebugInfo,
+    split_debuginfo: Option<String>,
+    debug_assertions: bool,
+    overflow_checks: bool,
+    rpath: bool,
+    incremental: bool,
+    panic: String,
+    #[serde(skip_serializing_if = "Vec::is_empty")] // remove when `rustflags` is stablized
+    rustflags: Vec<String>,
+}
+
+impl From<&Profile> for SbomProfile {
+    fn from(profile: &Profile) -> Self {
+        let rustflags = profile
+            .rustflags
+            .iter()
+            .map(|x| x.to_string())
+            .collect_vec();
+
+        Self {
+            name: profile.name.to_string(),
+            opt_level: profile.opt_level.to_string(),
+            lto: profile.lto,
+            codegen_backend: profile.codegen_backend.map(|x| x.to_string()),
+            codegen_units: profile.codegen_units.clone(),
+            debuginfo: profile.debuginfo.clone(),
+            split_debuginfo: profile.split_debuginfo.map(|x| x.to_string()),
+            debug_assertions: profile.debug_assertions,
+            overflow_checks: profile.overflow_checks,
+            rpath: profile.rpath,
+            incremental: profile.incremental,
+            panic: profile.panic.to_string(),
+            rustflags,
+        }
+    }
+}
+
+/// Describes a package dependency
+#[derive(Serialize, Clone, Debug)]
+struct SbomPackage {
+    package_id: PackageIdSpec,
+    package: String,
+    /// Package profile is given when it differs from the root level [`Profile`].
+    #[serde(skip_serializing_if = "Option::is_none")]
+    profile: Option<SbomProfile>,
+    version: Option<Version>,
+    features: Vec<String>,
+    build_type: SbomBuildType,
+    extern_crate_name: String,
+    /// Indices into the `packages` array
+    dependencies: Vec<usize>,
```

Comment:
Should we newtype the `usize`?  On one hand, it will make it easy to find all all `usize`s that fill those role.  On the other hand, it will make indexing more annoying.